android/app/src/main/java/org/mozilla/magnet/magnetapi/ApiPreferences.java

```diff
@@ -0,0 +1,106 @@
+package org.mozilla.magnet.magnetapi;
+
+import android.content.Context;
+
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.mozilla.magnet.api.Api;
+
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.Map;
+import java.util.HashMap;
+
+class ApiPreferences extends Api {
+
+    private static final String PREFERENCES_STORE_KEY = "preferences";
+    private static final String PREFERENCES_KEY = "pref_key";
+
+    private final Lock mWriteLock = new ReentrantLock();
+
+    ApiPreferences(Context aContext) {
+        super(aContext);
+    }
+
+    /**
+     * Get user's preferences.
+     *
+     */
+    @Override
+    public void get(String aPath, Callback aCallback) {
+        aCallback.resolve(getCache().getJsonObject(PREFERENCES_STORE_KEY));
+    }
+
+    /**
+     * Update or set a preference given a Map with structure:
+     * `{ "pref_key": "string", "value": "string" }`
+     */
+    @Override
+    public void post(String aPath, Object aData, Callback aCallback) {
+        if (!(aData instanceof Map)) {
+            aCallback.reject("Data must be a Map type");
+            return;
+        }
+
+        // Lock everything between getting the JSON object from the cache and
+        // writing it back - a synchronize on the method isn't enough as the
+        // resource is used in the delete method too.
+        mWriteLock.lock();
+
+        Map data = (Map) aData;
```

Comment:
'aData' is type Object, we need a 'Map', the check to ensure it's an instance of 'Map' happens before we acquire the lock.  Could move the cast up outside of the locked area too.